src/field/scrcmd_mystery_gift.c

```diff
@@ -0,0 +1,684 @@
+    (void)fieldSys;
+    SaveMGDataPtr_SetReceivedByIndex(SaveMGDataPtr_GetFirstGiftIndex());
+}
+
+BOOL ScrCmd_MysteryGift(ScriptContext* ctx) {
+    switch (ScriptReadHalfword(ctx)) {
+    case SCR_MG_BEGIN:
+        SaveMGDataPtr_Begin(ctx->fieldSystem->saveData, HEAP_ID_32);
+        break;
+    case SCR_MG_END:
+        SaveMGDataPtr_End(ctx->fieldSystem->saveData, FALSE);
+        break;
+    case SCR_MG_END2:
+        SaveMGDataPtr_End(ctx->fieldSystem->saveData, TRUE);
+        break;
+    case SCR_MG_HAS_GIFT: {
+        u16* ptr = ScriptGetVarPointer(ctx);
+        if (FieldSystem_GetTagOfNextMG(ctx->fieldSystem) != MG_TAG_INVALID) {
+            *ptr = TRUE;
+        } else {
+            *ptr = FALSE;
+        }
+        break;
+    }
+    case SCR_MG_GET_TYPE: {
+        u16* ptr = ScriptGetVarPointer(ctx);
+        *ptr = FieldSystem_GetTagOfNextMG(ctx->fieldSystem);
+        break;
+    }
+    case SCR_MG_CAN_RECEIVE: {
+        u16* ptr = ScriptGetVarPointer(ctx);
+        const struct ScriptMysteryGiftFuncs* pFunc = &sScriptMysteryGiftActionTable[FieldSystem_GetTagOfNextMG(ctx->fieldSystem) - 1];
+        *ptr = pFunc->check(ctx->fieldSystem, FieldSystem_GetDataOfNextMG(ctx->fieldSystem));
+        break;
+    }
+    case SCR_MG_RECEIVE:
+        const struct ScriptMysteryGiftFuncs* pFunc = &sScriptMysteryGiftActionTable[FieldSystem_GetTagOfNextMG(ctx->fieldSystem) - 1];
+        pFunc->give(ctx->fieldSystem, FieldSystem_GetDataOfNextMG(ctx->fieldSystem));
+        FieldSystem_SetQueuedMGReceived(ctx->fieldSystem);
+        break;
+    case SCR_MG_MESSAGE_RECEIVED: {
+        struct GetMysteryGiftGmmState gmmState;
+        const struct ScriptMysteryGiftFuncs* pFunc = &sScriptMysteryGiftActionTable[FieldSystem_GetTagOfNextMG(ctx->fieldSystem) - 1];
+        MessageFormat** pMsgFormat = FieldSysGetAttrAddr(ctx->fieldSystem, SCRIPTENV_MESSAGE_FORMAT);
+        u16* pMsgBank = ScriptGetVarPointer(ctx);
+        u16* pMsgNum = ScriptGetVarPointer(ctx);
+        FieldSystem_InitGetMysteryGiftGmmState(&gmmState, ctx->fieldSystem, *pMsgFormat, FieldSystem_GetDataOfNextMG(ctx->fieldSystem));
+        pFunc->messageSuccess(&gmmState, pMsgBank, pMsgNum);
+        break;
+    }
+    case SCR_MG_MESSAGE_FAILED: {
+        struct GetMysteryGiftGmmState gmmState;
+        const struct ScriptMysteryGiftFuncs* pFunc = &sScriptMysteryGiftActionTable[FieldSystem_GetTagOfNextMG(ctx->fieldSystem) - 1];
+        MessageFormat** pMsgFormat = FieldSysGetAttrAddr(ctx->fieldSystem, SCRIPTENV_MESSAGE_FORMAT);
+        u16* pMsgBank = ScriptGetVarPointer(ctx);
+        u16* pMsgNum = ScriptGetVarPointer(ctx);
+        FieldSystem_InitGetMysteryGiftGmmState(&gmmState, ctx->fieldSystem, *pMsgFormat, FieldSystem_GetDataOfNextMG(ctx->fieldSystem));
+        pFunc->messageFailure(&gmmState, pMsgBank, pMsgNum);
+        break;
+    }
+    }
+    return FALSE;
+}
+
+BOOL MGCheck_PartySpace(FieldSystem* fieldSys, MysteryGiftData* unused) {
```

Comment:
part of me thinks these should be named MysteryGift_Check_PartySpace

Reply:
this entire file is confused, it's partly scrcmd, partly mystery gift

Reply:
i do not share that part of you